Source/bmalloc/libpas/src/libpas/pas_stats.c

```diff
@@ -0,0 +1,466 @@
+            return buf->ptr;
+        free(buf->ptr);
+    }
+    buf->len = desired_len * 2;
+    buf->ptr = calloc(1, buf->len);
+    return buf->ptr;
+}
+
+static size_t pas_stats_malloc_info_bucket_idx_from_size(size_t size)
+{
+    size_t bucket_base;
+    size_t surplus_size;
+
+    if (size < PAS_STATS_MALLOC_INFO_HIGH_GRANULARITY_MAX_SIZE)
+        return size >> PAS_STATS_MALLOC_INFO_HIGH_GRANULARITY_SHIFT;
+    bucket_base = PAS_STATS_MALLOC_INFO_HIGH_GRANULARITY_BUCKETS;
+
+    if (size < PAS_STATS_MALLOC_INFO_MEDIUM_GRANULARITY_MAX_SIZE) {
+        surplus_size = size - PAS_STATS_MALLOC_INFO_HIGH_GRANULARITY_MAX_SIZE;
+        return bucket_base + (surplus_size >> PAS_STATS_MALLOC_INFO_MEDIUM_GRANULARITY_SHIFT);
+    }
+    bucket_base += PAS_STATS_MALLOC_INFO_MEDIUM_GRANULARITY_BUCKETS;
+
+    if (size < PAS_STATS_MALLOC_INFO_LOW_GRANULARITY_MAX_SIZE) {
+        surplus_size = size - PAS_STATS_MALLOC_INFO_MEDIUM_GRANULARITY_MAX_SIZE;
+        return bucket_base + (surplus_size >> PAS_STATS_MALLOC_INFO_LOW_GRANULARITY_SHIFT);
+    }
+    bucket_base += PAS_STATS_MALLOC_INFO_LOW_GRANULARITY_BUCKETS;
+
+    return bucket_base;
+}
+
+// Returns the minimum size for the bucket
+static size_t pas_stats_malloc_info_size_from_bucket_idx(size_t bucket_idx)
+{
+    const size_t high_granularity_bucket_bound = PAS_STATS_MALLOC_INFO_HIGH_GRANULARITY_BUCKETS;
+    const size_t medium_granularity_bucket_bound = high_granularity_bucket_bound
+        + PAS_STATS_MALLOC_INFO_MEDIUM_GRANULARITY_BUCKETS;
+    const size_t low_granularity_bucket_bound = medium_granularity_bucket_bound
+        + PAS_STATS_MALLOC_INFO_LOW_GRANULARITY_BUCKETS;
+
+    if (bucket_idx < high_granularity_bucket_bound)
+        return bucket_idx * PAS_STATS_MALLOC_INFO_HIGH_GRANULARITY;
+    if (bucket_idx < medium_granularity_bucket_bound)
+        return (bucket_idx - high_granularity_bucket_bound) * PAS_STATS_MALLOC_INFO_MEDIUM_GRANULARITY;
+    if (bucket_idx < low_granularity_bucket_bound)
+        return (bucket_idx - medium_granularity_bucket_bound) * PAS_STATS_MALLOC_INFO_LOW_GRANULARITY;
+    return PAS_STATS_MALLOC_INFO_LOW_GRANULARITY_MAX_SIZE;
+}
+
+void pas_stats_malloc_info_record(pas_stats_malloc_info_data* data, pas_stats_heap_type heap_type, size_t size, size_t count)
+{
+    PAS_TESTING_ASSERT(data);
+    PAS_TESTING_ASSERT(heap_type < pas_stats_heap_type_count);
+
+    size_t size_bucket_idx = pas_stats_malloc_info_bucket_idx_from_size(size);
+
+    pas_atomic_fetch_add_uint64_relaxed(&data->total_count, count);
+    pas_atomic_fetch_add_uint64_relaxed(&data->count_by_heap_type[heap_type], count);
+    pas_atomic_fetch_add_uint64_relaxed(&data->count_by_size[size_bucket_idx], count);
+}
+
+void pas_stats_log_all_enabled_stats(void)
+{
+    pas_lock_assert_held(&g_pas_stats_data.log_lock);
+
+    uint64_t log_time = current_time_ns();
+    if (!g_pas_stats_data.first_log_time_ns)
+        g_pas_stats_data.first_log_time_ns = current_time_ns();
```

Comment:
You're not using `first_log_time_ns` for anything other than to serve as an epoch.

Let's rename it to `start_time_ns` and initialize this once in `pas_stats_setup()` instead.  There's no value add in doing this check every time.